clients/cloud/rust/src/producer.rs

```diff
@@ -0,0 +1,64 @@
+/**
+ * Copyright 2019 Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+use futures::future::Future;
+use rdkafka::error::KafkaError;
+use rdkafka::message::OwnedMessage;
+use rdkafka::producer::{FutureProducer, FutureRecord};
+use std::boxed::Box;
+
+mod utils;
+
+fn log_produce_result(
+    topic: &str,
+    result: Result<(i32, i64), (KafkaError, OwnedMessage)>,
+) -> Result<(), ()> {
+    result
+        .and_then(|(p, o)| {
+            println!(
+                "Successfully produced record to topic {} partition [{}] @ offset {}",
+                topic, p, o
+            );
+            Ok(())
+        })
+        .map_err(|(err, _)| eprintln!("kafka error: {}", err))
+}
+
+fn main() -> Result<(), Box<std::error::Error>> {
+    let (topic, config) = utils::get_config()?;
+    let producer: FutureProducer = config.create()?;
+
+    let messages = (0..9)
+        .map(|msg| {
+            let value = msg.to_string();
+            println!("Preparing to produce record: {} {}", "alice", value);
+            producer.send(
+                FutureRecord::to(&topic)
+                    .payload(value.as_bytes())
+                    .key("alice"),
+                0,
+            )
+        })
+        .collect::<Vec<_>>();
+
+    for msg in messages {
+        msg.wait()
+            .map_err(|err| eprintln!("error producing message: {}", err))
+            .and_then(|result| log_produce_result(&topic, result))
+            .ok();
```

Comment:
Is this `.ok()` necessary?

Reply:
yes because i'm using `and_then` which returns a Result which means that there could be an error. We know there won't be because the result we're returning doesn't throw any error but we still want to make the compiler happy

Reply:
I'm not sure I follow, essentially we're going from ` Result<Result<(i32, i64), (KafkaError, OwnedMessage)>>` here to `Option<()>` but not consuming the option. That should be equivalent to `Result<Result<(i32, i64), (KafkaError, OwnedMessage)>> -> ()` (since we don't use the option. Calling `expect()` here would be fine since it's illegal for `log_produce_result` to return any error.

Reply:
i'm using `Result::and_then` not `Option::and_then` so a `Result` will be returned, not an `Option`

Reply:
Hmm, the `.ok()` here will consume the result and return an option.

Reply:
yea basically I'm saying, "even if the `and_then` has an error in the result (which it won't), it's okay anyways". That way, the compile won't complain about an unchecked result.